design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,381 @@
+
+### Global config
+
+The list of chunk containers is stored in a global `RepositoryConfig` object. There is a single configuration per repo, and all operations use the latest version of the configuration.
+
+The configuration is stored as a yaml file in `repo-root/config.yaml`, and it's written using PUT if-match to ensure writers are aware of the latest updates before overwriting.
+
+The global configuration solves also [other requirements](https://github.com/earth-mover/icechunk/issues/440) outside of this design doc.
+
+```rust
+pub struct RepositoryConfig {
+    pub inline_chunk_threshold_bytes: u16,
+    pub unsafe_overwrite_refs: bool,
+    // ... more fields
+
+    virtual_chunk_containers: Vec<VirtualChunkContainer>,
+}
+```
+
+The `RepositoryConfig` object acts as a default, read during repository instantiation, and overridable by users at runtime. Different fields of configuration may have different behaviors when overridden.
+
+A user can modify the `RepositoryConfig` for their current repo instance, but decide not to save the change to storage. This will be very common as users tweak the configuration during experimentation.
+
+### Setting virtual references
+
+As we mentioned, we want to make as little changes as possible to the way virtual chunks are written to Icechunk by the open source stack. The new `set_virtual_ref` function still takes a `location` url as the main data, adds an optional `checksum` and an optional `validate_containers` argument.
+
+If `validate_containers` is set to `True`, Icechunk will verify in `set_virtual_ref` that there is a `VirtualChunkContainer` with a `prefix` field  that is a prefix of the `location` url passed. An exception is raised if that's not the case and the reference is not stored.
+
+The following examples will work for `set_virtual_ref` called with the given `location` and `validate_containers=True`:
+
+* `location=s3://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=s3`.
+* `location=s3://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=s3://`.
+* `location=s3://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=s3://foo`.
+* `location=s3://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=s3://foo/bar.nc`.
+* `location=tigris://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=tigris`.
+* `location=models://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=models`.
+
+The following examples will now work for `set_virtual_ref` called with the given `location` and `validate_containers=True`:
+
+* `location=s3://foo/bar.nc` if there are no `VirtualChunkContainers`
+* `location=gcs://foo/bar.nc` if the only `VirtualChunkContainers` has `prefix=s3`
+
+But calling `set_virtual_ref` in the two last examples will still work with `validate_containers=False`.
```

Comment:
I'm not sure I understand the use case for `validate_containers=False`. When would this behavior be desirable?